Add Nested Custom Attributes EA (#2302)
braze-inc/braze-docs: _docs/_hidden/private_betas/nested_custom_attribute_support.md

---
nav_title: Nested Custom Attribute Support
permalink: "/nested_custom_attribute_support/"
hidden: true
---
<br>
{% alert note %}
Nested custom attribute support is currently in beta. Please contact your Braze account manager if you are interested in participating in the beta.
{% endalert %}

# Nested Custom Attribute Support

> Nested custom attribute support allows you to send objects as a new data type for custom attributes. This nested data allows you to create segments using information from a custom attribute object, and personalize your API-triggered messages using a custom attribute object and Liquid.

Reviewer: You can personalize all messages, no?
```suggestion
Nested custom attribute support allows you to send objects as a new data type for custom attributes. This nested data allows you to create segments using information from a custom attribute object, and personalize your messages using a custom attribute object and Liquid.
```

Objects can contain existing [data types][1], such as:

- Numbers
- Strings
- Booleans
- Arrays
- Other objects

## Limitations

- Available on custom attributes sent via API only, the Braze SDKs are not yet supported.
- Partners do not yet support nested custom attributes. Until this is supported, we recommend against using this feature with app groups that have partner integrations enabled.

> - Arrays do not currently support objects, coming soon. Looking for this feature? [Schedule a chat][4] with the product team.

Reviewer: I might phrase this as:
```suggestion
- Arrays of objects are not currently supported, but they are coming soon. Looking for this feature? [Schedule a chat][4] with the product team.
```

- Objects have a maximum size of 50KB.
- Key names and string values have a size limit of 255 characters.

## Usage Examples

### API Request Body

{% tabs local %}
{% tab Create %}

> Shown below is a `/users/track` example with a "Last Played Song" object. To capture the properties of the song, we'll send an API request that lists "last_played_song" as an object, and array of the nested properties of the song.

Reviewer: > and **array** of the nested properties of the song.

This seems inaccurate (emphasis mine).

Reviewer 2: Yep.  This one will need to be changed to remove array.

```

{
  "attributes": [
    {
      "external_id": "user_id",
      "_merge_objects" : true,
      "last_played_song" : {
          "year_released" : 1960
      }
    }
  ]
}
```

After the above request is received, the custom attribute object will now look like this:

```
"last_played_song" : {
    "song_name":"Solea",
    "artist_name" : "Miles Davis",
    "album_name": "Sketches of Spain",
    "genre" : "Jazz",
    "year_released" : 1960
}
```

{% alert warning %}
You must set `merge_objects` to true, or your objects will be overwritten. `_merge_objects` is false by default.
{% endalert %}

{% endtab %}
{% tab Delete %}
To delete a custom attribute object, send a POST to `users/track` with the custom attribute object set to `null`.

```
{
  "attributes": [
    {
      "external_id": "user_id",
      "last_played_song" : null
    }
  ]
}
```

{% endtab %}
{% endtabs %}

### Liquid Templating

> The Liquid templating example below shows how to reference the custom attribute object properties saved from the above API request and use them in your API-triggered Liquid messaging. 

Reviewer: ```suggestion
The Liquid templating example below shows how to reference the custom attribute object properties saved from the above API request and use them in your messaging. 
```

Use the `custom_attributes` personalization tag and dot notation to access properties on an object. Specify the name of the object, followed by a dot (period), followed by the property name.

{% raw %}
`{{custom_attribute.${last_played_song}.genre}}` - "Jazz"
<br> `{{custom_attribute.${last_played_song}.artist_name}}` - "Miles Davis"
{% endraw %}

![Last Played Push Notification Example][2]

### Segmentation

You can build Segments based on nested custom attributes to further target your users. To do so, filter your Segment based on the custom attribute object, then specify the property name and associated value you want to segment on.

> ![Last Played Song Segment][3]

Reviewer: It might be more instructive to use an example with one additional level of nesting. Here, and for templating as well.

Reviewer 2: Yep.  @bre-fitzgerald I've got an example of additional nesting here https://sweeney.appboy.com/engagement/campaigns/60dc6ccfd6e9b064b85ebc7d/5ce7f15abf5cea45ca07889a?locale=en&campaignName=Most%20Played%20Push&page=0
